Source/animation/types/animation_generators/TransformGenerator.swift

```diff
@@ -75,50 +79,49 @@ func addTransformAnimation(_ animation: BasicAnimation, sceneLayer: CALayer, ani
-        ty: CGFloat(transform.dy))
-}
-
-func transformAnimationByFunc(_ node: Node, valueFunc: (Double) -> Transform, duration: Double, offset: Double, fps: UInt) -> CAAnimation {
-
-    var transformValues = [CATransform3D]()
-    var timeValues = [Double]()
-
-    let step = 1.0 / (duration * Double(fps))
-    var dt = 0.0
-    var tValue = Array(stride(from: 0.0, to: 1.0, by: step))
-    tValue.append(1.0)
-    for t in tValue {
-
-        dt = t
-        if 1.0 - dt < step {
-            dt = 1.0
+func transformAnimationByFunc(_ animation: TransformAnimation, duration: Double, offset: Double, fps: UInt) -> CAAnimation {
+
+    let group = CAAnimationGroup()
+    group.duration = duration
+    group.fillMode = kCAFillModeForwards
+    group.isRemovedOnCompletion = false
+    
+    let fromTransform = animation.getVFunc()(0.0).toCG()
+    let toTransform = animation.getVFunc()(animation.autoreverses ? 0.5 : 1.0).toCG()
+    
+    var transformAnimation = CABasicAnimation(keyPath: "transform")
+    
+    if let spring = animation.spring {
+        switch spring {
+        case let .spring(mass, stiffness, damping, initialVelocity):
+            let springAnimation = CASpringAnimation(keyPath: "transform")
+            springAnimation.mass = mass
+            springAnimation.stiffness = stiffness
+            springAnimation.damping = damping
+            springAnimation.initialVelocity = initialVelocity
+            transformAnimation = springAnimation
+        default:
+            break
         }
-
-        timeValues.append(dt)
-
-        let view = nodesMap.getView(node)
-        let value = AnimationUtils.absoluteTransform(node, pos: valueFunc(offset + dt), view: view)
-        let cgValue = CATransform3DMakeAffineTransform(value.toCG())
-        transformValues.append(cgValue)
     }
-
-    let transformAnimation = CAKeyframeAnimation(keyPath: "transform")
+    
     transformAnimation.duration = duration
-    transformAnimation.values = transformValues
-    transformAnimation.keyTimes = timeValues as [NSNumber]?
+    transformAnimation.fromValue = NSValue(caTransform3D: CATransform3DMakeAffineTransform(fromTransform))
+    transformAnimation.toValue = NSValue(caTransform3D: CATransform3DMakeAffineTransform(toTransform))
     transformAnimation.fillMode = kCAFillModeForwards
     transformAnimation.isRemovedOnCompletion = false
-
-    return transformAnimation
-}
-
-func fixedAngle(_ angle: CGFloat) -> CGFloat {
-    return angle > -0.0000000000000000000000001 ? angle : CGFloat(2.0 * Double.pi) + angle
+    
+    group.animations = [transformAnimation]
+    
+    if let trajectory = animation.trajectory {
+        let pathAnimation = CAKeyframeAnimation(keyPath: "position")
+        pathAnimation.calculationMode = kCAAnimationPaced
+        pathAnimation.fillMode = kCAFillModeForwards
+        pathAnimation.isRemovedOnCompletion = false
+        pathAnimation.path = trajectory
+        
+        group.animations?.append(pathAnimation)
```

Comment:
Why `group.animation` is optional? Probably empty array should be fine.

Reply:
it is optional, and I am not the one who defined it that way)

Reply:
But you might be the one who make it right 😉

Reply:
it's CAAnimationGroup, the only thing I can is write a pleading/threatening letter to Apple. I can declare another array to use temporarily, if that will make it better